### PR TITLE
SINGLE_FILE + locateFile fix (#5952)

### PR DESCRIPTION
**Pull request by buu700**
Fix for #5952.

### CODE REVIEW

File: tests/test_browser.py [hunk 1]
```diff
@@ -3699,6 +3699,28 @@ def test_single_file_html(self):
     self.btest('emscripten_main_loop_setimmediate.cpp', '1', args=['-s', 'SINGLE_FILE=1', '-s', 'WASM=1', '-s', "BINARYEN_METHOD='native-wasm'"], also_proxied=True)
     assert os.path.exists('test.html') and not os.path.exists('test.js') and not os.path.exists('test.worker.js')
 
+  # Tests that SINGLE_FILE works as intended with locateFile
+  def test_single_file_locate_file(self):
+    open('src.cpp', 'w').write(self.with_report_result(open(path_from_root('tests', 'browser_test_hello_world.c')).read()))
+    Popen([PYTHON, EMCC, 'src.cpp', '-o', 'test.js', '-s', 'SINGLE_FILE=1', '-s', 'WASM=1', '-s', "BINARYEN_METHOD='native-wasm'"]).communicate()
```
Comment (kripken) [inline tests/test_browser.py]: Thanks! looks good, except on this line,

 * can replace `Popen` with the new `run_process`, which doesn't need `.communicate()` and also checks the return value.
 * can remove `-s BINARYEN_METHOD=native-wasm` as it is the default anyhow.

**Comment by kripken** [comment]
For code size and clarity, it might be nice to add a `isDataURI` method to replace all the `indexOf` calls, maybe?

**Comment by buu700** [comment]
Cool, just took care of those issues you mentioned.

**Comment by kripken** [comment]
Great, thanks!

**Comment by kripken** [comment]
This broke a few browser tests (that don't run here on CI), `browser.test_separate_asm` and `browser.test_pthread_sbrk`. Error is `isDataURI is not defined` so I guess we forgot something about what things HTML is ok to use.

**Comment by kripken** [comment]
Proposed fix in #5963

**Comment by buu700** [comment]
Oh, oops. The trouble I ran into was that base64Utils was being emitted after the the code in preamble that depended on it, and while the function `isDataURI` was hoisted the value or the var `dataURIPrefix` that it depended on wasn't, so as a quick fix I just moved `isDataURI` into preamble, as that seemed like a lower risk of breaking something.

So I guess the best fix would be to move that stuff back into base64utils and then emit base64utils before preamble. If that can't be done without breaking existing dependency relationships, then a few off-the-cuff ideas come to mind:

1. Make a new file containing `isDataURI` and load it in the HTML and before preamble.

2. Duplicate `dataURIPrefix` between the two functions that need it.

3. Make `dataURIPrefix` a function (for the hoisting).

It feels like a bit of a hack, but I think I like #3 out of those.

Oh, you beat me by a few seconds. Looking at your PR now.

**Comment by buu700** [comment]
Without testing your PR, as-is I think it would have the same problem I ran into (`dataURIPrefix` being undefined at the time that `integrateWasmJS` is executed), so it'd need to be printed before preamble.js, but otherwise looks good to me.

It also addresses the fact that `isDataURI` is needed even when `SUPPORT_BASE64_EMBEDDING` is false, when I guess I forgot about in my last comment.

**Pull request merged**